apps/web/src/pages/Users/UpdateUserPage/components/AddIndividualRoleDialog.tsx

```diff
@@ -0,0 +1,167 @@
+import React from "react";
+import { FormProvider, useForm } from "react-hook-form";
+import { useIntl } from "react-intl";
+import { PlusIcon } from "@heroicons/react/24/outline";
+
+import { Dialog, Button, IconButton } from "@gc-digital-talent/ui";
+import { MultiSelectField } from "@gc-digital-talent/forms";
+import { toast } from "@gc-digital-talent/toast";
+import {
+  commonMessages,
+  errorMessages,
+  getLocalizedName,
+} from "@gc-digital-talent/i18n";
+
+import {
+  Role,
+  UpdateUserAsAdminInput,
+  UpdateUserAsAdminMutation,
+  User,
+} from "~/api/generated";
+import { getFullNameHtml } from "~/utils/nameUtils";
+
+type FormValues = {
+  roles: Array<string>;
+};
+
+export type IndividualRoleSubmitData = Partial<UpdateUserAsAdminInput>;
+
+interface AddIndividualRoleDialogProps {
+  user: User;
+  availableRoles: Array<Role>;
+  onAddRoles: (
+    submitData: UpdateUserAsAdminInput,
+  ) => Promise<UpdateUserAsAdminMutation["updateUserAsAdmin"]>;
+}
+
+const AddIndividualRoleDialog = ({
+  user,
+  availableRoles,
+  onAddRoles,
+}: AddIndividualRoleDialogProps) => {
+  const intl = useIntl();
+  const [isOpen, setIsOpen] = React.useState<boolean>(false);
+  const userName = getFullNameHtml(user.firstName, user.lastName, intl);
+
+  const methods = useForm<FormValues>({
+    defaultValues: {
+      roles: [],
+    },
+  });
+
+  const {
+    handleSubmit,
+    formState: { isSubmitting },
+  } = methods;
+
+  const handleAddRoles = async (formValues: FormValues) => {
+    return onAddRoles({
+      roles: {
+        attach: {
+          roles: formValues.roles,
+        },
+      },
+    }).then(() => {
+      setIsOpen(false);
+      toast.success(
+        intl.formatMessage({
+          defaultMessage: "Role(s) added successfully",
+          id: "/17wgm",
+          description:
+            "Message displayed to user when one or more roles have been added to a user",
+        }),
+      );
+    });
+  };
+
+  const label = intl.formatMessage({
+    defaultMessage: "Add new role",
+    id: "2lNHxh",
+    description: "Label for the form to add a role to a user",
+  });
+
+  const roleOptions = availableRoles
+    .filter((role) => {
+      return !(
+        !role.isTeamBased ||
+        !user?.roleAssignments?.some(
+          (assignment) => assignment.role?.id === role.id,
+        )
+      );
```

Comment:
We only want roles that are (1) not team-based and (2) aren't in the user's current list of roleAssignments.

```suggestion
      return (
        !role.isTeamBased &&
        !user?.roleAssignments?.some(
          (assignment) => assignment.role?.id === role.id,
        )
      );
```

(This may have appeared correct bc the graphql query was missing isTeamBased on the query response)